### PR TITLE
[Flink]add range strategy for sort compaction

### PR DESCRIPTION
**Pull request by wg1026688210**
### Purpose

Improve the processing efficiency of sort compaction operations when there is data size skewness on range phase.

1. add  the  size based rangestrategy config.
2. add sample ratio config for per task and range num config.

### Tests

SortCompactActionForUnawareBucketITCase
RangeShuffleTest
InternalRowToSizeVisitorTest

### API and Format

add config  
``` --table_conf sort-compaction.range-strategy= SIZE ```

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/SortCompactActionForUnawareBucketITCase.java [hunk 1]
```diff
@@ -259,22 +261,26 @@ public void testRandomSuffixWorks() throws Exception {
     }
 
     private void zorder(List<String> columns) throws Exception {
+        String rangeStrategy = RANDOM.nextBoolean() ? "size" : "quantity";
         if (RANDOM.nextBoolean()) {
-            createAction("zorder", columns).run();
+            createAction("zorder", rangeStrategy, columns).run();
         } else {
-            callProcedure("zorder", columns);
+            callProcedure("zorder", rangeStrategy, columns);
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/SortCompactActionForUnawareBucketITCase.java:268]: I think we need not to add rangStrategy in the parameters. We can find anywhere in the sort process.

Comment (leaves12138) [reply paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/SortCompactActionForUnawareBucketITCase.java:268]: For example, we can find in SortUtils.sortStreamByKey, `Strategy strategy = table.coreOptions().sortRangeStrategy`

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/SortCompactActionForUnawareBucketITCase.java [hunk 2]
```diff
@@ -287,14 +293,21 @@ private SortCompactAction createAction(String orderStrategy, List<String> column
                 "--order_strategy",
                 orderStrategy,
                 "--order_by",
-                String.join(",", columns));
+                String.join(",", columns),
+                "--sort_conf range_strategy=" + rangeStrategy,
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/action/SortCompactActionForUnawareBucketITCase.java:297]: We can add this as table-config, when we do compaction, we just --table_conf sort.range.strategy = xxx. By this way, we can simplify code and keep compatibility with old code.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/shuffle/RangeShuffle.java [hunk 3]
```diff
@@ -486,4 +548,45 @@ public int get() {
...
+
+    @VisibleForTesting
+    static <T> T[] allocateRangeBaseSize(Tuple2<T, Integer>[] elements, int rangeNum) {
+        int elementsNum = elements.length;
+        long totalSize = Arrays.stream(elements).mapToLong(t -> (long) t.f1).sum();
+        long restSize = totalSize;
+        double targetSize = (totalSize) / (double) (rangeNum);
+
+        @SuppressWarnings("unchecked")
+        T[] range = (T[]) new Object[rangeNum];
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/shuffle/RangeShuffle.java:560]: why not rangeNum - 1?  If we separate a array num to 1000 part， we should set 999 point?

**Review by leaves12138 — COMMENTED** [review]
Thanks for @wg1026688210 a lot, just some comment of mine left. @JingsongLi please take a look at this pull request

**Comment by wg1026688210** [comment]
@leaves12138 @JingsongLi PTAL

**Review by leaves12138 — COMMENTED** [review]
Thanks for @wg1026688210 , comment above

File: paimon-common/src/main/java/org/apache/paimon/types/InternalRowToSizeVisitor.java [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    public BiFunction<DataGetters, Integer, Integer> visit(
+            LocalZonedTimestampType localZonedTimestampType) {
+        return (row, index) -> {
+            if (row.isNullAt(index)) {
+                return NULL_SIZE;
+            } else {
+                long x =
+                        row.getTimestamp(index, localZonedTimestampType.getPrecision())
+                                .getMillisecond();
+                return 8;
```
[the remark this answers is not in the document]
Comment (wg1026688210) [reply paimon-common/src/main/java/org/apache/paimon/types/InternalRowToSizeVisitor.java:209]: done

**Review by leaves12138 — APPROVED** [review]
+1

**Pull request merged**